ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -995,4 +1003,208 @@ void markStartTime() {
+      RestUtils.setArg(restRequest, InternalKeys.TARGET_CONTAINER_KEY, targetContainer, false);
+      logger.trace("Sets targetAccount={} targetContainer={} for restRequest={} ", targetAccount, targetContainer,
+          restRequest);
+    } catch (RestServiceException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new RestServiceException(
+          "Invalid blob id=" + blobIdStr + " during injecting account and container from blobId string",
+          RestServiceErrorCode.BadRequest);
+    }
+  }
+
+  /**
+   * Sanity check for {@link RestRequest}. This check ensures that the specified service id, account and container name,
+   * if they exist, should not be the same as the reserved values.
+   * @param restRequest The {@link RestRequest} to check.
+   * @throws RestServiceException if the specified service id, account or container name is set as system reserved value.
+   */
+  private void accountAndContainerSanityCheck(RestRequest restRequest) throws RestServiceException {
+    if (Account.UNKNOWN_ACCOUNT_NAME.equals(getTargetAccountNameFromHeader(restRequest)) || Account.UNKNOWN_ACCOUNT_NAME
+        .equals(getServiceId(restRequest))) {
+      throw new RestServiceException("Invalid account for putting blob", RestServiceErrorCode.InvalidAccount);
+    }
+    String targetContainerName = getTargetContainerNameFromHeader(restRequest);
+    if (Container.UNKNOWN_CONTAINER_NAME.equals(targetContainerName) || Container.UNKNOWN_PUBLIC_CONTAINER_NAME.equals(
+        targetContainerName) || Container.UNKNOWN_PRIVATE_CONTAINER_NAME.equals(targetContainerName)) {
+      throw new RestServiceException("Invalid container for putting blob", RestServiceErrorCode.InvalidContainer);
+    }
+  }
+
+  /**
+   * Sets target {@link Account} and {@link Container} objects in the {@link RestRequest}.
+   * @param restRequest The {@link RestRequest} to set.
+   * @param targetAccount The target {@link Account} to set.
+   * @param targetContainer The target {@link Container} to set.
+   */
+  private void setTargetAccountAndContainerInRestRequest(RestRequest restRequest, Account targetAccount,
+      Container targetContainer) {
+    RestUtils.setArg(restRequest, InternalKeys.TARGET_ACCOUNT_KEY, targetAccount, false);
+    RestUtils.setArg(restRequest, InternalKeys.TARGET_CONTAINER_KEY, targetContainer, false);
+    logger.trace("Sets targetAccount={} and targetContainer={} for restRequest={} ", targetAccount, targetContainer,
+        restRequest);
+  }
+
+  /**
+   * Ensures the {@link Account} matches the account name specified in the {@link RestRequest}, if it is specified.
+   * @param account The {@link Account} to ensure.
+   * @param restRequest The {@link RestRequest} to ensure.
+   * @throws RestServiceException if the {@link Account}'s name does not match the name specified in the {@link RestRequest}.
+   */
+  private void ensureAccountNameMatchOrThrow(Account account, RestRequest restRequest) throws RestServiceException {
```

Comment:
I think calling this just "ensureAccountNameMatch" is descriptive enough. Same for the method below.